### DISCUSSION TITLE
Global Ref for Dio Interceptors

### DISCUSSION THREAD

**Discussion by abikko** [rrousselGit/riverpod#2107] [Q&A] [answered]
Hello. I have a task for creating Dio instance with Custom Interceptors.
```

import 'package:flutter_riverpod/flutter_riverpod.dart';

final dioAuthInterceptorProvider = Provider((ref) => DioAuthInterceptor(ref));
class DioAuthInterceptor extends Interceptor {
  final Ref ref;
  DioAuthInterceptor(this.ref);

  @override
  void onError(DioError err, ErrorInterceptorHandler handler) {
    if (err.response?.data?['status'] == 1100) {
      ref.read(authNotifierProvider.notifier);
    }
    super.onError(err, handler);
  }
}

final dioTokenInterceptorProvider = Provider((ref) => DioTokenInterceptor(ref));
class DioTokenInterceptor extends Interceptor {
  final FetchAuthDataInLocalInteractor _fetchAuthDataInLocalInteractor = locator();
  final Ref ref;

  DioTokenInterceptor(this.ref);

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) async {
    options.headers.addAll(await buildTokenHeader());
    super.onRequest(options, handler);
  }

  Future<Map<String, dynamic>> buildTokenHeader() async {
    var authData = await ref.read(authNotifierProvider.notifier).fetchAuthData();
    if (authData != null) {
      return {'Authorization:': ' Bearer ${authData.accessToken}'};
    }
    return {};
  }
}
```

Also, I have DI modules, using [get_it](https://pub.dev/packages/get_it)

The question is - how I should to pass the Ref object into Interceptors? Without Provider.
Because I think it's not correct to connect DI with View package(riverpod).

**Comment by AhmedLSayed9** [comment, marked as answer]
You can't do that without passing `Ref`.
There's a way is to expose your main container globally and do `container.read` but that's unrecommended and an anti-pattern.

`Because I think it's not correct to connect DI with View package(riverpod).`
Riverpod Providers are a complete replacement for patterns like Singletons, Service Locators, Dependency Injection or InheritedWidgets.
So, the thing is you don't need Get_it when using riverpod, and mixing them can lead to issues. You might need Get_it for some other use-cases or scenarios but avoid mixing it with riverpod like for DI from presentation layer to data layer.
